config/v1/types_cluster_monitoring.go

```diff
@@ -0,0 +1,86 @@
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:internal
+// +openshift:api-approved.openshift.io=https://github.com/openshift/api/pull/1929
+// +openshift:file-pattern=cvoRunLevel=0000_10,operatorName=config-operator,operatorOrdering=01
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=clustermonitoring,scope=Cluster
+// +kubebuilder:subresource:status
+// +kubebuilder:metadata:annotations="description=Cluster Monitoring Operators configuration API"
+// +openshift:enable:FeatureGate=ClusterMonitoringConfig
+// ClusterMonitoring is the Schema for the Cluster Monitoring Operators API
+type ClusterMonitoring struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec holds user configuration for the Cluster Monitoring Operator
+	// +kubebuilder:validation:Required
+	Spec ClusterMonitoringSpec `json:"spec,omitempty"`
+	// status holds observed values from the cluster. They may not be overridden.
+	// +optional
+	Status ClusterMonitoringStatus `json:"status,omitempty"`
+}
+
+// MonitoringOperatorStatus defines the observed state of MonitoringOperator
+type ClusterMonitoringStatus struct {
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:internal
+type ClusterMonitoringList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ClusterMonitoring `json:"items"`
+}
+
+// ClusterMonitoringSpec defines the desired state of Cluster Monitoring Operator
+type ClusterMonitoringSpec struct {
+	// userDefined set the deployment mode for user-defined monitoring in addition to the default platform monitoring.
+	// +optional
+	UserDefined UserDefinedMonitoring `json:"userDefined,omitempty"`
+}
+
+// UserDefinedMonitoring config for user-defined projects.
+type UserDefinedMonitoring struct {
+	// mode defines the different configurations of UserDefinedMonitoring
+	// +kubebuilder:validation:Enum:="Disabled";"NamespaceIsolation"
+	Mode UserDefinedMode `json:"mode,omitempty"`
+}
+
+type UserDefinedMode string
+
+const (
+	// UserDefinedDisabled disables monitoring for user-defined projects. This restrics the default monitoring stack, installed in the openshift-monitoring project, to monitor only platform namespaces, which prevents any custom monitoring configurations or resources from being applied to user-defined namespaces.
+	UserDefinedDisabled UserDefinedMode = "Disabled"
```

Comment:
Does this supersede the presence of the user monitoring CRs that you plan to add? Could we validate somehow that disabling this disallows the creation of those CRs? And this cannot be disabled if user monitoring CRs are present?

Reply:
If this is disabled the user-defined projects are not enabled. The CR could be there just its not taking the config to create user-defined projects. As in the current ConfigMap we can switch the mode and activate user-defined projects.

Reply:
> The CR could be there just its not taking the config to create user-defined projects.

Would this be surprising to a user who has configured their monitoring, how do they understand that the monitoring is not currently enabled, is something going to tell them?

Reply:
When we add the user defined monitoring CRD, we'll communicate in its status if user defined monitoring is enabled or not.
I tihnk this is better then preventing its creation. The prevention path would raise the question of what to do with existing user defined objects when user defined is disabled. We wouldn't want to delete them (disbaling them might be temporary) so preventing addition (and editing) seems like not the best solution.

Reply:
What is the use case for being able to disable?

Could you prevent disabling, unless there were no user defined CRDs?

Reply:
The use case is that the decision whether to enable it lies with the cluster admin. The actual configuration can be done by a non-admin user. We think its useful for admins to be able to (perhaps temporarily) disable user defined monitoring without having to remove the config owned by a separate user group.

Reply:
If the cluster admin has disabled these resources, what signal will end users get to show that these are disabled? How will they be told that the configuration they have applied is now inert?